File: backend/schemas/analysis.py
"""
schemas/analysis.py
===================
Pydantic v2 request and response models for analysis job endpoints.
"""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class AnalyzeRequest(BaseModel):
    """Request body for POST /api/v1/analyze/{video_id}."""

    frame_sample_rate: int = Field(
        default=5,
        ge=1,
        le=30,
        description="Analyse every Nth frame (1 = every frame, 30 = every 30th).",
    )
    confidence_threshold: float = Field(
        default=0.45,
        ge=0.1,
        le=1.0,
        description="Minimum YOLO detection confidence to accept a bounding box.",
    )
    bandwidth_profile: Optional[str] = Field(
        default=None,
        description="Named bandwidth profile for simulation (see GET /bandwidth-profiles).",
    )
    run_comparison: bool = Field(
        default=False,
        description="If true, run all 3 strategies (Uniform ABR, Static ROI, SemanticStream) in parallel.",
    )


class RegionMetrics(BaseModel):
    """Quality metrics for a single priority region."""

    ssim: Optional[float]
    spqi: Optional[float]


class JobQueuedResponse(BaseModel):
    """Response returned immediately after queuing an analysis job."""

    job_id: str
    status: str = "queued"
    estimated_time_seconds: Optional[int]


class JobResultsResponse(BaseModel):
    """Full metrics payload for a completed analysis job."""

    job_id: str
    status: str
    progress_percent: float
    metrics: Optional[Dict[str, Any]] = None


class FrameAnnotationResponse(BaseModel):
    """Response for the single-frame annotation endpoint."""

    frame_base64: str
    priority_map_base64: str
    detections: List[Dict[str, Any]]

File: frontend/src/api/analysis.js
/**
 * api/analysis.js
 * Analysis job endpoints — queue a job and poll for results.
 */

import client from './client'

/**
 * Queue an analysis job for an uploaded video.
 * POST /api/v1/analyze/{video_id}
 * @param {string} videoId
 * @param {{ frame_sample_rate?: number, confidence_threshold?: number, bandwidth_profile?: string, run_comparison?: boolean }} config
 * @returns {Promise<{ job_id: string, status: string, estimated_time_seconds: number }>}
 */
export async function startAnalysis(videoId, config = {}) {
  return client.post(`/analyze/${videoId}`, {
    frame_sample_rate: config.frameSampleRate ?? 5,
    confidence_threshold: config.confidenceThreshold ?? 0.45,
    bandwidth_profile: config.bandwidthProfile ?? null,
    run_comparison: config.runComparison ?? false,
  })
}

/**
 * Get status and metrics for an analysis job.
 * GET /api/v1/results/{job_id}
 * @param {string} jobId
 * @returns {Promise<{ job_id: string, status: string, progress_percent: number, metrics: object|null }>}
 */
export async function getResults(jobId) {
  return client.get(`/results/${jobId}`)
}
